### PR TITLE
Implement C\is_sorted and C\is_sorted_by

### PR DESCRIPTION
**Pull request by lexidor**

### CODE REVIEW

**Comment by fredemmott** [comment]
Just seeing is_sorted here? :)

**Comment by lexidor** [comment]
We could/should restict `Tv` with an `as num`. Anybody who wants to keep the wishy washy, anything goes nature, with the documentation?

**Comment by lexidor** [comment]
> Just seeing is_sorted here? :)

You're just too fast. I still gotta do that one. I should have made this a draft PR instead.

**Comment by fredemmott** [comment]
> We could/should restict Tv with an as num. Anybody who wants to keep the wishy washy, anything goes nature, with the documentation?

I don't think that should be done here: it's technically correct, and Vec\sort() is technically wrong - I'd rather have consistency, and fix both at the same time later.

**Comment by fredemmott** [comment]
`as num` is also only needed for the default comparator

**Comment by lexidor** [comment]
> `as num` is also only needed for the default comparator

That means we either need `Tv (as num if $comparator is null)`, but shoehorning that in the typesystem if not going to be an easy task. If would not want to have two functions, where one checks nums (with an optional cmp) and one that checks anything with a required cmp.

**Comment by lexidor** [comment]
Also, all files are `.php`, but mine are (right now) `.hack`. Should I stay consistent with the rest?

**Comment by fredemmott** [comment]
> `as num`

Yeah, I'd rather leave it fixmed as it is at the moment - and as it /should/ be in Vec\sort but we're not picking up the problem there

> `.php`

.php please; for repositories with an internal source of truth (pushed by facebook-github-bot , and with 'fbshipit-source-id' in commit messages), we can't currently have `.hack` files.

**Comment by lexidor** [comment]
### Feature is completed



> Just seeing is_sorted here? :)

This is when I should have undrafted this PR.

**Comment by lexidor** [comment]
I failed to pass the linter? This project does not have hhast-lint.json at its base. Running hhast-lint with this config:
```JS
{
  "roots": [ "src/" ],
  "builtinLinters": "default"
}
```
causes this diff:
```
git diff | wc
272     805    9465
```

I don't know what linters are turned on in the test suite, so I can't pass it, without linting the entire codebase at the strictest possible level.

This would also make changes to the File and OS namespaces, which are being worked on by someone else rn. I don't want to be causing merge conflicts.

**Comment by jjergus** [comment]
The Facebook-internal linter is unrelated to hhast-lint (unfortunately -- I'm trying to figure out if we can share more code between them).

It's complaining about trailing whitespace. You can probably run `git show --ws-error-highlight=all` to see where the trailing whitespace is.

File: tests/c/COrderTest.php [hunk 1]
```diff
@@ -0,0 +1,197 @@
...
+      tuple(vec[1, 0], false),
+      // Containers with two strictly equal elements are sorted
+      tuple(vec[100, 100], true),
+      // Containers with two weakly equal elements are sorted
+      tuple(vec[100., 100], true),
+      // Containers with n sorted elements are sorted
+      tuple(Vec\range(100, 1000, 11), true),
+      // Containers with n reversed elements are not sorted
+      tuple(Vec\range(1000, 100, 11), false),
+      // Containers with alphabetically sorted strings are sorted
```
Comment (jjergus) [inline tests/c/COrderTest.php]: sorry, there's one more here :/

File: tests/c/COrderTest.php [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+      tuple(vec[100, 100], true),
+      // Containers with two weakly equal elements are sorted
+      tuple(vec[100., 100], true),
+      // Containers with n sorted elements are sorted
+      tuple(Vec\range(100, 1000, 11), true),
+      // Containers with n reversed elements are not sorted
+      tuple(Vec\range(1000, 100, 11), false),
+      // Containers with alphabetically sorted strings are sorted 
+      tuple(vec['aaa', 'bbb', 'c', 'ca', 'ccc'], true),
+      // Containers with lexiographically sorted strings are not sorted
```
Comment (jjergus) [inline tests/c/COrderTest.php]: and here...I think that's all of them

**Comment by lexidor** [comment]
@jjergus Regexing for `/ \n/` finds nothing, but the linter is not happy. What is causing it to fail?

**Comment by jjergus** [comment]
As far as I can tell all the other errors are because of differences between our internal tooling and the open-source equivalents (HackTest, expect...), so probably nothing you can do about those. I can figure out how to fix them before committing. I'm also trying to get our internal and OSS tooling closer together, to minimize issues like this, but it will take a while before I can make any visible progress on that :(

**Comment by jjergus** [comment]
here's a list of FB-internal issues I had to deal with (mostly as a note to myself so that I know what I should work on in the future):

- add typehints to lambda function arguments where they can't be inferred (we have a lint rule requiring this)
- correct combination of @oss-enable and @oss-disable on use clauses
- DataProviders must be static
- DataProvider return type must match test function argument type
- internally gettype is in a "PHP" namespace so using it directly requires FIXMEs
- internal expect only has toEqual, not toBeSame

**Pull request closed without merging**